### PR TITLE
Sort Tool Search Result by ranking

### PR DESCRIPTION
**Pull request by juleengraham**

### CODE REVIEW

**Comment by dannon** [comment]
@juleengraham Thanks for opening the PR for this.  I'll go ahead and check it out, but the commit history here looks like something went wrong.  Can you clean that up?  Let me know if you need help.

**Comment by juleengraham** [comment]
Address issue #3276 and part of  #8772

**Comment by juleengraham** [comment]
@galaxybot test this

**Comment by juleengraham** [comment]
@galaxybot test this

**Comment by martenson** [comment]
Hi @juleengraham. This PR does not work for me. For any search string I am getting

```
vue.runtime.esm.js?2b0e:1888 TypeError: Cannot read property 'forEach' of undefined
    at eval (utilities.js?b271:51)
    at Array.map (<anonymous>)
    at filterTools (utilities.js?b271:50)
    at VueComponent.categories (ToolBox.vue?02cd:92)
```

This is because when you run `category.elems.forEach(...`and one of the "categories" is a tool section label, it has no `elems`


Other observations: 

- the filtering should only trigger (and the buttons display) when there is at least one result available.
- it would be nice to have a test for this



As a side note it would be nice to unify the vocabulary used here and consistently use tool panel "section" instead of "category". I believe that is what Galaxy uses in configuration and elsewhere.

Also using active forms of verbs in docs is better than the passive forms. e.g.:

>//Filters out to only display categories that have tools that were in results

`Filter out categories without tools in results`

**Comment by martenson** [comment]
@juleengraham thank you for the quick turnaround. The filtering now works well.

I suggest to make not showing section names the default (and still offer a button to show them). 

But more importantly I believe that in case where we do not show sections we should default to pure ordering by the ranking. This is what was [suggested](https://github.com/galaxyproject/galaxy/issues/8772#issuecomment-542269850) in the issue and I expect that it will make the results more consistent.

Currently when the #2 search hit is in the last section of the tool panel, it will be displayed in the very end of the results (no matter if with sections or without). A better way would be to display it second when no sections are visible, and in the section when they are. I am not sure whether sections should be reordered to respect the rankings or not. Maybe we can leave that part as is and decide later.

**Comment by juleengraham** [comment]
@martenson Sure, I can make sorting without sections the default.
Re the sorting, it is currently set to have tools listed in the order that they rank. Is that not happening on your end? Do you have an example?

Screen Shot 2020-04-28 at 7 42 43 PM
Screen Shot 2020-04-28 at 7 42 52 PM

**Comment by martenson** [comment]
@juleengraham sure, this is from my local instance, with fairly normal toolpanel

Galaxy
Galaxy

**Comment by martenson** [comment]
I found one more thing. When clearing the results with an X there is a following console error:

```
TypeError: Cannot read property 'length' of undefined
    at VueComponent.onQuery (ToolBox.vue?02cd:131)
    at invokeWithErrorHandling (vue.runtime.esm.js?2b0e:1854)
    at VueComponent.invoker (vue.runtime.esm.js?2b0e:2179)
    at invokeWithErrorHandling (vue.runtime.esm.js?2b0e:1854)
    at VueComponent.Vue.$emit (vue.runtime.esm.js?2b0e:3888)
    at VueComponent.checkQuery (ToolSearch.vue?1716:39)
```

**Comment by juleengraham** [comment]
@martenson 
`[Vue warn]: Invalid prop: type check failed for prop "category". Expected Object, got Number with value 0.`
This error might have been fixed with the latest commit but I wasn't able to reproduce it so let me know if it's still appearing and if the button still appears when there are no results.

File: client/galaxy/scripts/components/Panels/ToolBox.vue [hunk 1]
```diff
@@ -128,7 +115,7 @@ export default {
     methods: {
         onQuery(query) {
             this.query = query;
-            this.queryFilter = (query.length >= 3) ? query : null
+            this.queryFilter = (query && query.length) >= 3 ? query : null;
```
Comment (martenson) [inline client/galaxy/scripts/components/Panels/ToolBox.vue]: This probably won't work since you're comparing a bool to an int.

**Comment by martenson** [comment]
@juleengraham I can still see it (two different ones actually)
Galaxy

File: client/galaxy/scripts/components/Panels/ToolBox.vue [hunk 2]
```diff
@@ -106,6 +115,7 @@ export default {
     methods: {
         onQuery(query) {
             this.query = query;
+            this.queryFilter = (query && (query.length >= 3)) ? query : null;
```
Comment (guerler) [inline client/galaxy/scripts/components/Panels/ToolBox.vue]: Do we need this? The `ToolSearch` component has `minQueryLength` already set to `3`.

File: client/galaxy/scripts/components/Panels/ToolBox.vue [hunk 3]
```diff
@@ -79,15 +88,19 @@ export default {
...
+                return filterTools(this.toolbox, this.results);
+            }
         },
         isUser() {
             const Galaxy = getGalaxyInstance();
             return !!(Galaxy.user && Galaxy.user.id);
         },
         workflows() {
-            return [
+            this.workflow = [
```
Comment (dannon) [inline client/galaxy/scripts/components/Panels/ToolBox.vue]: Computed properties should never have side effects (manipulating state)

**Comment by martenson** [comment]
@juleengraham thanks for all the fixes, it looks very nice, great job!

p.s. please ping me next time, just a rebase does not create notification so I'd have to check periodically which does not work for me atm.

**Pull request merged**